internal/service/kendra/index.go

```diff
@@ -0,0 +1,1204 @@
+										ValidateFunc: validation.All(
+											validation.StringLenBetween(1, 2048),
+											validation.StringMatch(
+												regexp.MustCompile(`^(https?|ftp|file):\/\/([^\s]*)`),
+												"Must be valid URL",
+											),
+										),
+									},
+									"user_name_attribute_field": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.StringLenBetween(1, 100),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"tags":     tftags.TagsSchema(),
+			"tags_all": tftags.TagsSchemaComputed(),
+		},
+	}
+}
+
+func resourceIndexCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).KendraConn
+	defaultTagsConfig := meta.(*conns.AWSClient).DefaultTagsConfig
+	tags := defaultTagsConfig.MergeTags(tftags.New(d.Get("tags").(map[string]interface{})))
+
+	name := d.Get("name").(string)
+
+	input := &kendra.CreateIndexInput{
+		ClientToken: aws.String(resource.UniqueId()),
+		Name:        aws.String(name),
+		RoleArn:     aws.String(d.Get("role_arn").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("edition"); ok {
+		input.Edition = types.IndexEdition(v.(string))
+	}
+
+	if v, ok := d.GetOk("server_side_encryption_configuration"); ok {
+		input.ServerSideEncryptionConfiguration = expandServerSideEncryptionConfiguration(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("user_context_policy"); ok {
+		input.UserContextPolicy = types.UserContextPolicy(v.(string))
+	}
+
+	if v, ok := d.GetOk("user_group_resolution_configuration"); ok {
+		input.UserGroupResolutionConfiguration = expandUserGroupResolutionConfiguration(v.([]interface{}))
+	}
+
+	if v, ok := d.GetOk("user_token_configurations"); ok {
+		input.UserTokenConfigurations = expandUserTokenConfigurations(v.([]interface{}))
+	}
+
+	if len(tags) > 0 {
+		input.Tags = Tags(tags.IgnoreAWS())
+	}
+
+	log.Printf("[DEBUG] Creating Kendra Index %#v", input)
+	output, err := conn.CreateIndex(ctx, input)
+
+	if err != nil {
+		return diag.Errorf("error creating Kendra Index (%s): %s", name, err)
+	}
+
+	if output == nil {
+		return diag.Errorf("error creating Kendra Index (%s): empty output", name)
+	}
+
+	d.SetId(aws.ToString(output.Id))
+
+	// arn needed to update tags
+	arn := arn.ARN{
+		Partition: meta.(*conns.AWSClient).Partition,
+		Service:   "kendra",
+		Region:    meta.(*conns.AWSClient).Region,
+		AccountID: meta.(*conns.AWSClient).AccountID,
+		Resource:  fmt.Sprintf("index/%s", d.Id()),
+	}.String()
+
+	d.Set("arn", arn)
+
+	// waiter since the status changes from CREATING to either ACTIVE or FAILED
+	if _, err := waitIndexCreated(ctx, conn, d.Id(), d.Timeout(schema.TimeoutCreate)); err != nil {
+		return diag.Errorf("error waiting for Index (%s) creation: %s", d.Id(), err)
+	}
+
+	return resourceIndexUpdate(ctx, d, meta)
+}
+
```

Comment:
since we won't need to update tags on create as they're passed directly in the `CreateIndex `method above, what if we only do an update when the `capacity_units` param is present as that is the only param we really need to do the update for e.g.

```suggestion
	// waiter since the status changes from CREATING to either ACTIVE or FAILED
	if _, err := waitIndexCreated(ctx, conn, d.Id(), d.Timeout(schema.TimeoutCreate)); err != nil {
		return diag.Errorf("error waiting for Index (%s) creation: %s", d.Id(), err)
	}
	
	if v, ok := d.GetOk("capacity_units"); ok && len(v.([]interface{})) > 0 && v.([]interface{})[0] != nil {
	        return resourceIndexUpdate(ctx, d, meta)
	}

	return resourceIndexRead(ctx, d, meta)
}

```

Reply:
Oh there's also the `document_metadata_configuration_updates` that we have to update. Shall I move it up as well to group it with the `capacity_units`?